cmd/analyzer/analyzer.go

```diff
@@ -239,12 +241,37 @@ func NewService(cfg *config.AnalysisConfig) (*Service, error) {
 func (a *Service) Start() {
 	a.logger.Info("starting analysis service")
 
+	var ctx context.Context
+	ctx, a.cancelAnalyzers = context.WithCancel(context.Background())
+
+	// trap Ctrl+C and call cancel on the context
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, os.Interrupt)
+	defer func() {
+		signal.Stop(signalChan) // Stop catching Ctrl+C signals.
+		a.Shutdown()            // Shut down gracefully.
+	}()
+
+	go func() {
+		select {
+		case <-signalChan:
+			// We received the first Ctrl+C; try to shut down gracefully.
+			a.Shutdown()
```

Comment:
the caller of Start, either runAnalyzer or func in rootMain calls .Shutdown. should only have to return, rather than putting these a.Shutdown calls in various places. are we able to return?

Reply:
oh are we not able to return because we don't cancel the context until something calls Shutdown thus calling cancelAnalyzers? seems like it would be more natural to receive the signal and only call cancelAnalyzers. no a.Shutdown(). and let us return normally

Reply:
that might also let us extract this into a helper that only creates a ctx that cancels on ctrl+c

Reply:
I don't fully understand this comment. Is your main concern that the caller of `Start()` now needs to be careful to call `Shutdown()`, therefore a clean shutdown is not guaranteed? Not sure what kind of returning you had in mind; where and why would the callers want to return? 

This might be relevant: The analysis `Service` (which owns/controls all the analyzers; the other service we have is the API service) was originally designed with explicit cleanup, i.e. with Start+Shutdown. So the Service's `Shutdown()` is always called when appropriate, e.g. [here](https://github.com/oasisprotocol/oasis-indexer/blob/dc3834f3132e79a9e9b07ce61d41b0dea6b81898/cmd/analyzer/analyzer.go#L70-L72) (and in another spot; there are only two calls to Start() anyway).

Reply:
yup, referring to that. i.e. it's (or was previously at least) up to the thing using the analyzer `Service` (1) to call its Start and (2) to call its Shutdown.

my understanding of this change is that Shutdown is now something that sometimes gets called by the service itself inside Start, sometimes from outside. in fact, if I'm following this right, both?

I feel it's a smaller change and less complicated to cancel the context and let the subordinate analyzers finish, let the Start return, and let the thing using the analyzer `Service` call Shutdown as before

Reply:
> Shutdown is now something that sometimes gets called by the service itself inside Start, sometimes from outside

That was indeed funky. Thank you for noticing.

Your observation caused me to think about the code more, and I noticed more smells that I then refactored away:
- The `Shutdown()` method of the `Service` _had_ to be called by the caller of `Service::Start()`. Then it makes more sense to just call `defer Shutdown()` from `Start()` and not burden the caller. We do so now.
- The code introduced by this PR repeated that sin (Start+Shutdown) at the Analyzer level. I changed it now to also use `defer`.
- Several `Shutdown()` methods throughout the codebase were IMO misleadingly named. They did not actively cause anything to shut down; they performed resource cleanup and assumed that any active processing had completed. I renamed them appropriately, IIRC to either `cleanup()` or `Close()`.

I also refactored the code around this comment thread so it no longer waits for ctrl+C in a separate goroutine; I find the new approach much easier to grok.

I added all these changes as a separate commit, should you prefer to view them that way, but I think it makes more sense to diff against main.

Reply:
oo looking good

little asymmetry, but don't let this distract you: initializing those resources is in a public function `NewMain` etc, while cleaning them up is in a private method `cleanup`. once you make a service, you have no way out other than to Start it and let Start clean things up when it's done